docs/design/provisioning.md

```diff
@@ -0,0 +1,223 @@
+* Allow a single means of configuration and consumption of storage (i.e, make it easy for admins and end users)
+* Allow easy update of storage config without server restart
+
+Summary implementation:
+
+Administrators use ConfigMap in a namespace they control to configure many provisioners.  Each provisioner creates 1 distinct kind and class of volume.  Using ConfigMap allows configuration to be stored in the API, which allows for easy update, watches, etc.
+
+A PersistentVolumeSelector added to PersistentVolumeClaim would first attempt to match labels on available PVs before looking for a provisioner with a matching label set, allowing "storage classes" to be defined for both manually created and dynamically provisioned volumes.
+
+
+## Using ConfigMap
+
+### Admin namespace
+
+Provisioners will be configured using `ConfigMap`.  `ConfigMap` allows configuration to be stored in the API server, which allows dynamic updates, querying, watches, etc.  The alternative is file config, which requires the same machinery as the API server (versioning, watches, querying, etc), or a new 1st class API object.  `ConfigMap` perfectly suits the requirements of this proposal without requiring a new API type.
+
+`ConfigMap` objects live in a namespace, so the administrator will create a namespace specifically to hold provisioner configuration.  This "system namespace" may contain many types of ConfigMaps, not all of which are provisioners, so a specific label must be applied to provisioners to identify them.
+
+The  namespace and label are required by kube-controller-manager so that the provisioner controller can query the correct namespace and find the provisioner configs.
+
+Admins will pass the namespace and label to the controller via CLI flag:
+
+```
+CLI flags:
+
+--pv-provisioner-namespace=foobar
+--pv-provisioner-label="type=provisioner-config
+
+example usage:
+
+kubectl --namespace=system-namespace get configmap -l type=provisioner-config
+
+```
+
+
+### Provisioner Configuration
+
+Use case:  Configure provisioners for GCE Persistent Disks in two zones (east and west) and two volume types (standard and solid state).  The admin chooses to use the term "storage-class" as a stand-in for volume type (speed) and uses the labels "gold" and "silver" for fast and slow.
+
+A distinct set of labels on a provisioner is the "storage class".  The use case requires 4 provisioners: gold+east, gold+west, silver+east, silver+west.   Each provisioner's "parameters" map would contain the values for volume type and zone.  The plugin uses the parameters to create that one type of PersistentVolume.
+
+One configuration is shown below.  The remainder can be found [here](provisioning-examples.json).
+
+```
+  {
+    "kind": "ConfigMap",
+    "apiVersion": "v1",
+    "metadata": {
+      "name": "gold-east",
+      "namespace": "kubernetes.io/provisioners",
+      "labels": {
+        "type": "provisioner-config",
+        "storage-class": "gold",
+        "zone": "us-east"
```

Comment:
We're not matching by configMap.Name.  We're matching labels in the claim's PVSelector with configMap.Labels.  This allows us to have "gold" mean one thing, instead of requiring the user to add "gold-east" and "gold-west" to their selector. Using separate labels ("storage-class"="gold" and "zone"="east") allows that.